src/bluez/adapter.rs

```diff
@@ -0,0 +1,117 @@
+use futures::stream::{Stream, StreamExt};
+use log::warn;
+use std::pin::Pin;
+
+#[derive(Clone, Debug)]
+pub struct Adapter {
+    session: BluetoothSession,
+    adapter: AdapterId,
+}
+
+impl Adapter {
+    pub(crate) fn new(session: BluetoothSession, adapter: AdapterId) -> Self {
+        Self { session, adapter }
+    }
+}
+
+#[async_trait]
+impl Central<Peripheral> for Adapter {
+    async fn events(&self) -> Result<Pin<Box<dyn Stream<Item = CentralEvent>>>> {
+        let events = self.session.event_stream().await?;
+        let session = self.session.clone();
+        Ok(Box::pin(events.filter_map(move |event| {
+            central_event(event, session.clone())
+        })))
+    }
+
+    async fn start_scan(&self) -> Result<()> {
+        self.session.start_discovery().await?;
+        Ok(())
+    }
+
+    async fn stop_scan(&self) -> Result<()> {
+        self.session.stop_discovery().await?;
+        Ok(())
+    }
+
+    async fn peripherals(&self) -> Result<Vec<Peripheral>> {
+        let devices = self.session.get_devices().await?;
+        Ok(devices
+            .into_iter()
+            .map(|device| Peripheral::new(self.session.clone(), device))
+            .collect())
+    }
+
+    async fn peripheral(&self, address: BDAddr) -> Result<Peripheral> {
+        let devices = self.session.get_devices().await?;
+        devices
+            .into_iter()
+            .find_map(|device| {
+                if BDAddr::from(&device.mac_address) == address {
+                    Some(Peripheral::new(self.session.clone(), device))
+                } else {
+                    None
+                }
+            })
+            .ok_or(Error::DeviceNotFound)
+    }
+}
+
+impl From<BluetoothError> for Error {
+    fn from(error: BluetoothError) -> Self {
+        Error::Other(error.to_string())
+    }
+}
+
+async fn central_event(event: BluetoothEvent, session: BluetoothSession) -> Option<CentralEvent> {
+    match event {
+        BluetoothEvent::Device {
+            id,
+            event: DeviceEvent::Discovered,
+        } => {
+            let device = session.get_device_info(&id).await.ok()?;
+            Some(CentralEvent::DeviceDiscovered((&device.mac_address).into()))
+        }
+        BluetoothEvent::Device {
+            id,
+            event: DeviceEvent::Connected { connected },
+        } => {
+            let device = session.get_device_info(&id).await.ok()?;
+            if connected {
+                Some(CentralEvent::DeviceConnected((&device.mac_address).into()))
+            } else {
+                Some(CentralEvent::DeviceDisconnected(
+                    (&device.mac_address).into(),
+                ))
+            }
+        }
+        BluetoothEvent::Device {
+            id,
+            event: DeviceEvent::RSSI { rssi: _ },
+        } => {
+            let device = session.get_device_info(&id).await.ok()?;
+            Some(CentralEvent::DeviceUpdated((&device.mac_address).into()))
+        }
```

Comment:
The old behaviour of the BlueZ implementation would always emit a `CentralEvent::DeviceDiscovred` first before a `CentralEvent::DeviceUpdated` if a device was unknown to _btleplug_ at the time. This was done with a `AdapterManager<Peripheral>`. 
This implementation does not keep track of the state of known devices, thus emits `CentralEvent::DeviceUpdated` with no corresponding `CentralEvent::DeviceDiscovered`. This makes sense, since the device would already be known to BlueZ at the time, but not to _btleplug_. 

Since this is unspecified behaviour for _btleplug_, I'm not sure if it's required to address before merging, or if we should start writing up a spec and then have the platform implementations follow

Reply:
I've added some code to make the event stream always start with `DeviceDiscovered` events for every device known at that point. What do you think?

Reply:
> if we should start writing up a spec and then have the platform implementations follow

You mean the step that this whole project was waiting on in the first place? :3